docs/public/guidelines/languages/go.md

```diff
@@ -0,0 +1,222 @@
+
+//// tab | VS Code / Cursor
+See the [wiki](https://github.com/golang/vscode-go/wiki/features#format-and-organize-imports) for more details.
+
+```json
+{
+  "gopls": {
+    "formatting.local": "github.com/DataDog/datadog-agent"
+  }
+}
+```
+////
+
+//// tab | Vim
+Configure [vim-go](https://github.com/fatih/vim-go) as follows.
+
+```vim
+let g:go_fmt_options = {
+\ 'goimports': '-local github.com/DataDog/datadog-agent',
+\ }
+```
+////
+///
+
+## Public APIs
+
+Go [supports](https://go.dev/doc/modules/layout#package-or-command-with-supporting-packages) the use of private `internal` packages to control the public API of a Go module. This prevents packages from being used outside of the parent module, which is useful for decoupling different parts of our codebase.
+
+When adding new code, carefully consider what API is exported by both taking care of what symbols are uppercase and by making judicious use of `internal` directories.
+
+1. When in doubt, prefer hiding public APIs as it's much easier to refactor code to expose something that was private than doing it the other way around.
+1. When refactoring a struct into its own package, try moving it into its own `internal` directory if possible (see example below).
+1. If the directory you are editing already has `internal` directories, try to move code to these `internal` directories instead of creating new ones.
+1. When creating new `internal` directories, try to use the deepest possible `internal` directory to limit the packages that can import yours. For example, if making `a/b/c/d` internal, consider moving it to `a/b/c/internal/d` instead of `a/internal/b/c/d`. With the first path, code from `a/b` won't be able to access the `d` package, while it could with the second path.
+
+/// details | Example
+Sometimes one wants to hide private fields of a struct from other code in the same package to enforce a particular code invariant. In this case, the struct should be moved to a different folder within the same package **making this folder `internal`**.
+
+Consider a module named `example.com` where you want to move `exampleStruct` from the `a` package to a subfolder to hide its private fields from `a`'s code.
+
+Before the refactor, the code will look like this:
+
+//// tab | :octicons-file-code-16: example.com/a/code.go
+```go
+package a
+
+type exampleStruct struct {
+    // some fields
+}
+
+func doSomethingWithExampleStruct(e exampleStruct) {
+    // some code goes here ...
+}
+```
+////
+
+After the refactor, you should move `exampleStruct` to an `a/internal/b` directory:
+
+//// tab | :octicons-file-code-16: example.com/a/internal/b/examplestruct.go
+```go
+package b
+
+type ExampleStruct struct {
+    // some fields
+}
+```
+////
+
+and import this package from `a/code.go`:
+
+//// tab | :octicons-file-code-16: example.com/a/code.go
+```go
+package a
+
+import (
+    "example.com/a/internal/b"
+)
+
+func doSomethingWithExampleStruct(e b.exampleStruct) {
```

Comment:
To keep the example consistent, to access the struct, we should use `b.ExampleStruct`